omp/matrix/fft_kernels.cpp

```diff
@@ -0,0 +1,334 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/fft_kernels.hpp"
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/base/allocator.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+/**
+ * @brief The FFT matrix format namespace.
+ * @ref Fft
+ * @ingroup fft
+ */
+namespace fft {
+
+
+template <typename InValueType, typename OutValueType>
+void bfly(const matrix::Dense<InValueType> *b, matrix::Dense<OutValueType> *x,
+          int64 lo, int64 hi, OutValueType root)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        auto lo_val = b->at(lo, rhs);
+        auto hi_val = b->at(hi, rhs);
+        x->at(lo, rhs) = lo_val + hi_val;
+        x->at(hi, rhs) = (lo_val - hi_val) * root;
+    }
+}
+
+
+template <typename ValueType>
+void bfly(matrix::Dense<ValueType> *x, int64 lo, int64 hi, ValueType root)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        auto lo_val = x->at(lo, rhs);
+        auto hi_val = x->at(hi, rhs);
+        x->at(lo, rhs) = lo_val + hi_val;
+        x->at(hi, rhs) = (lo_val - hi_val) * root;
+    }
+}
+
+
+template <typename ValueType>
+void bit_rev_swap(matrix::Dense<ValueType> *x, int64 i, int64 rev_i)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        if (i < rev_i) {
+            std::swap(x->at(i, rhs), x->at(rev_i, rhs));
+        }
+    }
+}
+
+
+int64 bit_rev(int64 i, int64 size)
+{
+    int64 rev{};
+    for (int64 fwd = 1, bwd = size / 2; fwd < size; fwd *= 2, bwd /= 2) {
+        rev |= ((i / fwd) & 1) * bwd;
```

Comment:
I am not sure whether compiler will optimize it but it can be replaced by `>> log(fwd)` and `<< log(bwd)` if know the log(size)

Reply:
yes, that would be possible as well, probably there are some smarter approaches in *Hacker's Delight*, but since this is only a single permutation step without much performance implications, I went for the straightforward approach first.